### ISSUE TITLE
Proper field paths in error messages

### ISSUE THREAD

**Issue by ctapobep** [qala-io/assert-reflection-equals#15]
When pointing to a field that's different between 2 objects, we need to specify a path to the fields in term of the field names: `mother.dog.id` Instead of `Person.Dog.id`.

**Comment by ctapobep** [comment]
1. Add the root class(?)
2. Show which element isn't equal in the arrays
3. Performance with primitives - should we push to the Deque fields even before we checked for equality? Or maybe we should first check if they are equal and only if not - add the field name to the Deque.

**Comment by gorshkovayulia** [comment]
Could you please take a look at this example and confirm that this is ok for us (or suggest an alternative approach):
```
Expected: Insect<id=0, bacteria=Bacteria<id=1, insect=Insect<id=5, bacteria=null, size=5.55>, size=2.88>, size=1.55>
Actual: Insect<id=0, bacteria=Bacteria<id=1, insect=Insect<id=6, bacteria=null, size=6.55>, size=2.88>, size=1.55>

--- Found following differences ---
Insect.bacteria.insect.size: expected: 5.55, actual: 6.55
Insect.bacteria.insect.id: expected: 5, actual: 6

--- Difference detail tree ---
 expected: Insect<id=0, bacteria=Bacteria<id=1, insect=Insect<id=5, bacteria=null, size=5.55>, size=2.88>, size=1.55>
 actual: Insect<id=0, bacteria=Bacteria<id=1, insect=Insect<id=6, bacteria=null, size=6.55>, size=2.88>, size=1.55>

bacteria expected: Bacteria<id=1, insect=Insect<id=5, bacteria=null, size=5.55>, size=2.88>
bacteria   actual: Bacteria<id=1, insect=Insect<id=6, bacteria=null, size=6.55>, size=2.88>

bacteria.insect expected: Insect<id=5, bacteria=null, size=5.55>
bacteria.insect   actual: Insect<id=6, bacteria=null, size=6.55>

bacteria.insect.size expected: 5.55
bacteria.insect.size   actual: 6.55

bacteria.insect.id expected: 5
bacteria.insect.id   actual: 6
```

**Comment by ctapobep** [comment]
I think we can fail-fast. I don't think we need diff's for _each_ field that doesn't match the expected result. Just for the 1st one.

**Comment by gorshkovayulia** [comment]
Confirm that for arrays we show only the first mismatch (not the full list as in Unitils lib, I need to rewrite this part as well)
```
--- Found the following differences ---

charArray expected: [a, b, c]
charArray actual: [d, e, f]

charArray[0] expected: a
charArray[0] actual: d
```

**Comment by ctapobep** [comment]
For arrays it's fine to list elements of both arrays. Basically "which field is different" - it's the arrays, so here they are.

As an _additional_ info it's probably nice to have this this data too:

```
charArray[0] expected: a
charArray[0] actual: d
```
But here we don't want to list _all_ the different elements - just the first diff. So the output that you showed is exactly what I'd expect - no more, no less.

**Comment by gorshkovayulia** [comment]
As for the first part of error message: in case bidirectional relationship (e.g. Insect -> Bacteria -> Insect), what would you expect to see? (Unitils lib doesn't support this case -> we're getting OOM)

I'm thinking about smth like that (don't show internal object if we've already showed it before). Is it clear or not? Or it looks like internal object is empty... ->
`Expected: Insect<id=1, bacteria=Bacteria<id=1, insect=Insect<>, size=1.88f>, size=1.55f>`

**Comment by ctapobep** [comment]
Yeah, we can't go into the same object twice. We can make it a little more clear for our users:

```
Expected: Insect<id=1, bacteria=Bacteria<id=1, insect=<BEEN HERE, NOT GOING INSIDE AGAIN>, size=1.88f>, size=1.55f>
```

**Comment by gorshkovayulia** [comment]
Yep, I haven't actually reproduced this problem - found my mistake.
So, the last question about "Difference detail tree" section: I think we should omit this section - I believe that this section is actually unnecessary if we're going to show only 1st error.

Instead, lets just show the problem object entirely (only if it's internal object):
`Expected: Insect<id=0, bacteria=Bacteria<id=1, insect=Insect<id=5, bacteria=null, size=5.55>, size=2.88>, size=1.55>
Actual: Insect<id=0, bacteria=Bacteria<id=1, insect=Insect<id=6, bacteria=null, size=6.55>, size=2.88>, size=1.55>

--- Found the following differences ---
expected: Insect<id=5, bacteria=null, size=5.55>
actual: Insect<id=6, bacteria=null, size=6.55>

bacteria.insect.size: expected: 5.55, actual: 6.55
bacteria.insect.id: expected: 5, actual: 6
`
Also note that I've deleted the class name on the way to the problem value

**Comment by ctapobep** [comment]
Yes, this looks good. I'd change the order to first see the fields that differ, and then the objects themselves. Maybe something like this:

```
Expected: Insect<id=0, bacteria=Bacteria<id=1, insect=Insect<id=5, bacteria=null, size=5.55>, size=2.88>, size=1.55>
Actual:      Insect<id=0, bacteria=Bacteria<id=1, insect=Insect<id=6, bacteria=null, size=6.55>, size=2.88>, size=1.55>

--- Fields that differed ---
bacteria.insect.id expected: 5
bacteria.insect.id actual:       6

--- Objects that differed ---
expected: Insect<id=5, bacteria=null, size=5.55>
actual:      Insect<id=6, bacteria=null, size=6.55>
```

PS: when writing down "actual:", let's put some extra spaces or a `\t` so that actual and expected values/objects were aligned - it's easier to compare them this way.
